Embed the pinned marketing-name table instead of reading a runtime data file.

--- a/amdgpu/amdgpu_asic_id.c
+++ b/amdgpu/amdgpu_asic_id.c
@@ -25,6 +24,0 @@
-// secure_getenv requires _GNU_SOURCE
-#ifndef _GNU_SOURCE
-#define _GNU_SOURCE
-#endif
-
-#include <ctype.h>
@@ -35 +28,0 @@
-#include <unistd.h>
@@ -41,70 +34 @@
-
-static int parse_one_line(struct amdgpu_device *dev, const char *line)
-{
-	char *buf, *saveptr;
-	char *s_did;
-	uint32_t did;
-	char *s_rid;
-	uint32_t rid;
-	char *s_name;
-	char *endptr;
-	int r = -EINVAL;
-
-	/* ignore empty line and commented line */
-	if (strlen(line) == 0 || line[0] == '#')
-		return -EAGAIN;
-
-	buf = strdup(line);
-	if (!buf)
-		return -ENOMEM;
-
-	/* device id */
-	s_did = strtok_r(buf, ",", &saveptr);
-	if (!s_did)
-		goto out;
-
-	did = strtol(s_did, &endptr, 16);
-	if (*endptr)
-		goto out;
-
-	if (did != dev->info.asic_id) {
-		r = -EAGAIN;
-		goto out;
-	}
-
-	/* revision id */
-	s_rid = strtok_r(NULL, ",", &saveptr);
-	if (!s_rid)
-		goto out;
-
-	rid = strtol(s_rid, &endptr, 16);
-	if (*endptr)
-		goto out;
-
-	if (rid != dev->info.pci_rev_id) {
-		r = -EAGAIN;
-		goto out;
-	}
-
-	/* marketing name */
-	s_name = strtok_r(NULL, ",", &saveptr);
-	if (!s_name)
-		goto out;
-
-	/* trim leading whitespaces or tabs */
-	while (isblank(*s_name))
-		s_name++;
-	if (strlen(s_name) == 0)
-		goto out;
-
-	dev->marketing_name = strdup(s_name);
-	if (dev->marketing_name)
-		r = 0;
-	else
-		r = -ENOMEM;
-
-out:
-	free(buf);
-
-	return r;
-}
+#include "ailo_amdgpu_ids.h"
@@ -168,106 +91,0 @@
-#if HAVE_SECURE_GETENV
-static char *join_path(const char *dir, const char *file) {
-	size_t dir_len = strlen(dir);
-	size_t file_len = strlen(file);
-	char *full_path = NULL;
-
-	int need_slash = ((dir_len > 0) && (dir[dir_len - 1] != '/'));
-	size_t total_len = dir_len + (need_slash ? 1 : 0) + file_len + 1; // +1 for null terminator
-
-	if (dir_len == 0) {
-		return strdup(file);
-	}
-
-	full_path = malloc(total_len);
-	if (!full_path) {
-		return NULL; // Memory allocation failed
-	}
-
-	strcpy(full_path, dir);
-	if (need_slash) {
-		full_path[dir_len] = '/';
-		dir_len++;
-	}
-	strcpy(full_path + dir_len, file);
-
-	return full_path;
-}
-
-static char **split_env_var(const char *env_var_content)
-{
-	char **ret = NULL;
-	char *dup_env_val;
-	int elements = 1;
-	int index = 1;
-
-	if (!env_var_content || env_var_content[0] == '\0')
-		return NULL;
-
-	for(char *p = (char *)env_var_content; *p; p++) {
-		if (*p == ':')
-			elements++;
-	}
-
-	dup_env_val = strdup(env_var_content);
-	if (!dup_env_val) {
-		return NULL;
-	}
-	ret = malloc(sizeof(char *) * (elements + 1));
-	ret[0] = dup_env_val;
-	for(char *p = (char *)dup_env_val; *p; p++) {
-		if (*p == ':') {
-			*p = 0;
-			ret[index++] = p + 1;
-		}
-	}
-	ret[index] = NULL; // ensure that the last element in the array is NULL
-	return ret;
-}
-
-static void split_env_var_free(char **split_var)
-{
-	if (split_var) {
-		// remember that the first element also points to the whole duplicated string,
-		// which was modified in place by replacing ':' with '\0' characters
-		free(split_var[0]);
-		free(split_var);
-	}
-}
-
-static char *find_asic_id_table(void)
-{
-	// first check the paths in AMDGPU_ASIC_ID_TABLE_PATHS environment variable
-	const char *amdgpu_asic_id_table_paths = secure_getenv("AMDGPU_ASIC_ID_TABLE_PATHS");
-	const char *file_name = NULL;
-	char *found_path = NULL;
-	char **paths = NULL;
-
-	if (!amdgpu_asic_id_table_paths)
-		return NULL;
-
-	// extract the file name from AMDGPU_ASIC_ID_TABLE
-	file_name = strrchr(AMDGPU_ASIC_ID_TABLE, '/');
-	if (!file_name)
-		return NULL;
-	file_name++; // skip the '/'
-
-	paths = split_env_var(amdgpu_asic_id_table_paths);
-	if (!paths)
-		return NULL;
-
-	// for each path, join with file_name and check if it exists
-	for (int i = 0; paths[i] != NULL; i++) {
-		char *full_path = join_path(paths[i], file_name);
-		if (!full_path) {
-			continue;
-		}
-		if (access(full_path, R_OK) == 0) {
-			found_path = full_path;
-			break;
-		}
-	}
-	split_env_var_free(paths);
-	return found_path;
-}
-#endif
-
@@ -276,47 +94,4 @@
-	FILE *fp;
-	char *line = NULL;
-	size_t len = 0;
-	ssize_t n;
-	int line_num = 1;
-	int r = 0;
-
-	char *amdgpu_asic_id_table_path = NULL;
-#if HAVE_SECURE_GETENV
-	// if this system lacks secure_getenv(), don't allow extra paths
-	// for security reasons.
-	amdgpu_asic_id_table_path = find_asic_id_table();
-#endif
-	// if not found, use the default AMDGPU_ASIC_ID_TABLE path
-	if (!amdgpu_asic_id_table_path)
-		amdgpu_asic_id_table_path = strdup(AMDGPU_ASIC_ID_TABLE);
-
-	fp = fopen(amdgpu_asic_id_table_path, "r");
-	if (!fp) {
-		fprintf(stderr, "%s: %s\n", amdgpu_asic_id_table_path,
-			strerror(errno));
-		goto get_cpu;
-	}
-
-	/* 1st valid line is file version */
-	while ((n = getline(&line, &len, fp)) != -1) {
-		/* trim trailing newline */
-		if (line[n - 1] == '\n')
-			line[n - 1] = '\0';
-
-		/* ignore empty line and commented line */
-		if (strlen(line) == 0 || line[0] == '#') {
-			line_num++;
-			continue;
-		}
-
-		drmMsg("%s version: %s\n", amdgpu_asic_id_table_path, line);
-		break;
-	}
-
-	while ((n = getline(&line, &len, fp)) != -1) {
-		/* trim trailing newline */
-		if (line[n - 1] == '\n')
-			line[n - 1] = '\0';
-
-		r = parse_one_line(dev, line);
-		if (r != -EAGAIN)
+	for (size_t i = 0; i < sizeof(ailo_amdgpu_ids) / sizeof(ailo_amdgpu_ids[0]); i++) {
+		if (dev->info.asic_id == ailo_amdgpu_ids[i].device &&
+		    dev->info.pci_rev_id == ailo_amdgpu_ids[i].revision) {
+			dev->marketing_name = strdup(ailo_amdgpu_ids[i].name);
@@ -324,10 +99 @@
-
-		line_num++;
-	}
-
-	if (r == -EINVAL) {
-		fprintf(stderr, "Invalid format: %s: line %d: %s\n",
-			amdgpu_asic_id_table_path, line_num, line);
-	} else if (r && r != -EAGAIN) {
-		fprintf(stderr, "%s: Cannot parse ASIC IDs: %s\n",
-			__func__, strerror(-r));
+		}
@@ -335,6 +100,0 @@
-
-	free(line);
-	fclose(fp);
-
-get_cpu:
-	free(amdgpu_asic_id_table_path);
@@ -342 +102 @@
-	    dev->marketing_name == NULL) {
+	    dev->marketing_name == NULL)
@@ -344 +103,0 @@
-	}
--- a/amdgpu/meson.build
+++ b/amdgpu/meson.build
@@ -22 +22,9 @@
-datadir_amdgpu = join_paths(get_option('prefix'), get_option('datadir'), 'libdrm')
+ailo_ids_script = files('ailo_embed_ids.cmake')
+ailo_ids = custom_target(
+  'ailo_amdgpu_ids',
+  input : '../data/amdgpu.ids',
+  output : 'ailo_amdgpu_ids.h',
+  command : [find_program('cmake', native : true), '-DINPUT=@INPUT@',
+             '-DOUTPUT=@OUTPUT@', '-P', ailo_ids_script],
+  depend_files : ailo_ids_script,
+)
@@ -31,0 +40 @@
+    ailo_ids,
@@ -36 +44,0 @@
-    '-DAMDGPU_ASIC_ID_TABLE="@0@"'.format(join_paths(datadir_amdgpu, 'amdgpu.ids')),
--- a/meson.build
+++ b/meson.build
@@ -359 +358,0 @@
-subdir('data')
